### Урок 6. Жизненный цикл A/b теста. Практика. Расчёт объёма и определение сроков тестирования

Цель домашнего задания:

Научиться рассчитывать необходимый объем выборки с помощью онлайн-калькуляторов и ручной проверки с использованием Python.

Чтобы сделать работу используйте Jupyter Notebook или в Google Colab. Вы также можете выполнять эту работу в Google-документах, предоставив доступ для редактирования


Что нужно сделать:

1. Решите следующие задания сначала с помощью онлайн-калькулятора, а затем самостоятельно с помощью Python:

— Вы работаете с рассылкой предложений перед “черной пятницей”, доля открываемых писем, исходя из заголовка письма в рассылках, обычно составляет 15% (конверсия). Вам хочется увеличить конверсию открываемых писем до 25% с помощью смайликов. Рассчитайте необходимый размер выборки. Сделай вывод. (В вашей базе рассылок около 40000 человек)

— И вот вы теперь тестируете две версии рассылок. И заметили, что в одной версии конверсия 19%, а в другой 24%. Всего у вас 100 человек в рассылке. Является ли результат статистически значимым? Если нет, то сколько должно было человек, чтобы результат был уже статистически значим?

— Вас повысили до старшего аналитика. Ваша первая задача на новой позиции, определить какой поставщик товаров (A или B) позволяет вам получить наибольшую оценку качества. У вас две группы оценок для A и B. Среднее группы A = 45, дисперсия группы A = 4, средние группы B = 40, дисперсия группы B = 3. Является ли группа А лучшей, если опираться на статистическую значимость? А если бы обе дисперсии были равны 1?

— Вы планируете эксперимент по способам расстановки продуктов на полке. Вы предполагаете, что некоторые пары продуктов надо класть вместе. В качестве метрики у вас средний чек покупок. Вы проводите эксперимент в 1000 магазинах. Средний чек в группе A = 870 рублей, дисперсия этого чека - 80 рублей. Какое минимальное значение метрики надо получить для группы B, чтобы после всей 1000 экспериментов с каждым магазином мы могли с уверенностью сказать, что тестируемая гипотеза в группе B верна?

2. В решении укажите результат, предложенный онлайн-калькулятором, а также полное решение задач в Python. Помимо кода, пропишите к каждой задаче, решенной в Python, комментарий, поясняющий полученные вами результаты.

3. Сравните результаты онлайн-калькулятора и самостоятельного решения и пропишите вывод по каждой задаче.


Рекомендации по выполнению:

— Рекомендованный калькулятор: [Sample Size Calculator](https://www.evanmiller.org/ab-testing/sample-size.html) Можно использовать любой другой, но в таком случае прикрепите ссылку на используемый калькулятор.
— Нужно использовать статистические методы и критерии.

### Task 1

Вы работаете с рассылкой предложений перед “черной пятницей”, доля открываемых писем, исходя из заголовка письма в рассылках, обычно составляет 15% (конверсия). Вам хочется увеличить конверсию открываемых писем до 25% с помощью смайликов. Рассчитайте необходимый размер выборки. Сделай вывод. (В вашей базе рассылок около 40000 человек)

In [1]:
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
def find_n(pa, p0, alpha = 0.05, beta = 0.2):
    za = stats.norm().ppf(1 - alpha)
    zb = stats.norm().ppf(1 - beta)
    sn = (np.sqrt(p0 * (1 - p0)) * za + np.sqrt(pa * (1 - pa)) * zb) / (pa - p0)
    return sn ** 2

In [3]:
alpha = 0.05
beta = 0.2

BCR = 0.15
MDE = 0.25
find_n(BCR, MDE, alpha=0.05, beta=0.2)

102.56856602139405

Требуемый размер выборки 103 человека.

С помощью онлайн-калькулятора при увеличении конверсии на 10% мы получаем требуемый размер выборки 214 человек. Поскольку на калькуляторе значение получилось больше, будем ориентироваться на него.

### Task 2

И вот вы теперь тестируете две версии рассылок. И заметили, что в одной версии конверсия 19%, а в другой 24%. Всего у вас 100 человек в рассылке. Является ли результат статистически значимым? Если нет, то сколько должно было человек, чтобы результат был уже статистически значим?

In [4]:
from scipy.stats import norm

n1 = 100
n2 = 100
p1 = 0.19
p2 = 0.24

p_pool = (n1 * p1 + n2 * p2) / (n1 + n2)
se = np.sqrt(p_pool * (1 - p_pool) * (1 / n1 + 1 / n2))
z = (p1 - p2) / se
p_value = 2 * (1 - norm.cdf(abs(z)))

p_value

0.3894588156343448

Поскольку p_value (0,38) > 0.05 мы не можем отвергнуть нулевую гипотезу. Это означает, что разница в коэффициентах конверсии между двумя версиями не является статистически значимой

In [5]:
BCR = 0.19
MDE = 0.24
find_n(BCR, MDE, alpha=0.05, beta=0.2)

426.5530238125583

Требуемый размер выборки для того, чтобы результат был статистически значимым, 427 человек.

С помощью онлайн-калькулятора мы получаем требуемый размер выборки 993 человека. Поскольку на калькуляторе значение получилось больше, будем ориентироваться на него.

### Task 3

Вас повысили до старшего аналитика. Ваша первая задача на новой позиции, определить какой поставщик товаров (A или B) позволяет вам получить наибольшую оценку качества. У вас две группы оценок для A и B. Среднее группы A = 45, дисперсия группы A = 4, средние группы B = 40, дисперсия группы B = 3. Является ли группа А лучшей, если опираться на статистическую значимость? А если бы обе дисперсии были равны 1?

In [6]:
# Предположим, что размеры выборок равны
n_A = 30
n_B = 30

std_A = np.sqrt(4)
std_B = np.sqrt(3)

mean_A = 45
mean_B = 40

t_statistic, p_value = stats.ttest_ind_from_stats(
    mean1=mean_A, std1=std_A, nobs1=n_A,
    mean2=mean_B, std2=std_B, nobs2=n_B,
    equal_var=False
)

print(f"p-значение: {p_value}")

if p_value < 0.05:
    print("Группа A статистически лучше группы B.")
else:
    print("Нет статистически значимых различий между группами.")

p-значение: 1.0813919362386744e-14
Группа A статистически лучше группы B.


Согласно онлайн калькулятору вывод такой же: среднее значение по выборке А больше

In [7]:
# Дисперсии равны
std_equal = np.sqrt(1)

t_statistic_equal, p_value_equal = stats.ttest_ind_from_stats(
    mean1=mean_A, std1=std_equal, nobs1=n_A,
    mean2=mean_B, std2=std_equal, nobs2=n_B,
    equal_var=True)

print(f"p-значение: {p_value_equal}")

if p_value_equal < 0.05:
    print("Группа A статистически лучше группы B.")
else:
    print("Нет статистически значимых различий между группами.")

p-значение: 5.3670133030694715e-27
Группа A статистически лучше группы B.


Согласно онлайн калькулятору вывод такой же: среднее значение по выборке А больше.

Вывод: В обоих случаях мы можем сделать вывод, что группа А является лучшим поставщиком с точки зрения оценки качества, основываясь на статистической значимости.

### Task 4

Вы планируете эксперимент по способам расстановки продуктов на полке. Вы предполагаете, что некоторые пары продуктов надо класть вместе. В качестве метрики у вас средний чек покупок. Вы проводите эксперимент в 1000 магазинах. Средний чек в группе A = 870 рублей, дисперсия этого чека - 80 рублей. Какое минимальное значение метрики надо получить для группы B, чтобы после всей 1000 экспериментов с каждым магазином мы могли с уверенностью сказать, что тестируемая гипотеза в группе B верна?

In [8]:
mean_A = 870
var_A = 80
n = 1000

std_A = np.sqrt(var_A)

standard_error = std_A / np.sqrt(n)

alpha = 0.05

t_critical = stats.t.ppf(1 - alpha / 2, df=n - 1)

# Минимальное значение среднего чека в группе B

min_mean_B = mean_A + t_critical * standard_error

print(f"Минимальное значение среднего чека в группе B для статистической значимости: {min_mean_B:.2f} рублей")

Минимальное значение среднего чека в группе B для статистической значимости: 870.56 рублей


Согласно онлайн-калькулятору минимальное значение среднего чека в группе B - 875.3 рубля, что выше нашего расчетного.